app/javascript/packages/analytics/index.spec.ts

```diff
@@ -26,44 +40,139 @@ describe('trackEvent', () => {
     });
 
     context('no payload', () => {
-      it('fetches and resolves to undefined', async () => {
+      it('calls sendBeacon and resolves to undefined', async () => {
         const result = await trackEvent('name');
 
         expect(result).to.be.undefined();
-        expect(global.fetch).to.have.been.calledWith(
-          endpoint,
-          sandbox.match({
-            body: '{"event":"name"}',
-            headers: { 'Content-Type': 'application/json' },
-            method: 'POST',
-          }),
-        );
+
+        expect(global.navigator.sendBeacon).to.have.been.calledOnce();
+
+        const [actualEndpoint, data] = (global.navigator.sendBeacon as SinonStub).firstCall.args;
+        expect(actualEndpoint).to.eql(endpoint);
+        expect(data).to.have.property('type').eql('application/json');
+
+        expect(await blobTextContents(data)).to.eql('{"event":"name"}');
+      });
+
+      it('does not fall back to fetch', async () => {
+        await trackEvent('name');
+        expect(global.fetch).not.to.have.been.called();
       });
     });
 
     context('payload', () => {
-      it('fetches and resolves to undefined', async () => {
+      it('calls sendBeacon and resolves to undefined', async () => {
         const result = await trackEvent('name', { foo: 'bar' });
 
         expect(result).to.be.undefined();
-        expect(global.fetch).to.have.been.calledWith(
-          endpoint,
-          sandbox.match({
-            body: '{"event":"name","payload":{"foo":"bar"}}',
-            headers: { 'Content-Type': 'application/json' },
-            method: 'POST',
-          }),
-        );
+
+        expect(global.navigator.sendBeacon).to.have.been.calledOnce();
+
+        const [actualEndpoint, data] = (global.navigator.sendBeacon as SinonStub).firstCall.args;
```

Comment:
how come the old code didn't need to cast `global.fetch` but this needs to cast `global.navigator.sendBeacon`?

Reply:
The fetch version wraps fetch in an `expect()`:

```typescript
          expect(global.fetch).to.have.been.calledWith(
            endpoint,
            sandbox.match({
              body: '{"event":"name","payload":{"foo":"bar"}}',
              headers: { 'Content-Type': 'application/json' },
              method: 'POST',
            }),
          );
```

`sandbox.match` doesn't work well with Blobs, because Blobs (at least in the implementation we are running for our JS tests) need a separate `FileReader` instance to actually examine the text content they contain.

Reply:
ooo weird, thx!